feat(edit-content): enhance text area component with Monaco editor support

oidacra opened the pull request:
This commit introduces significant updates to the component, enabling dual editing modes: plain text and code editing via the Monaco editor. Key changes include:

- Refactored the HTML structure to accommodate a dropdown for selecting the editor type and a language variable selector.
- Implemented a new for the Monaco editor integration, including custom styling and configuration options.
- Added utility functions for detecting the appropriate editor based on content type.
- Updated SCSS styles for improved layout and consistency across the text area component.

### Proposed Changes
This pull request includes significant changes to the `dot-edit-content-text-area` and `dot-edit-content-wysiwyg-field` components to enhance their functionality and improve the user experience. The most important changes include the addition of a new editor mode for text areas, refactoring of the WYSIWYG field component, and updates to the supporting styles and utilities.

Enhancements to `dot-edit-content-text-area` component:

* Added support for switching between plain text and Monaco code editor modes, including the necessary dropdown options and event handling (`core-web/libs/edit-content/src/lib/fields/dot-edit-content-text-area/dot-edit-content-text-area.component.html`, `core-web/libs/edit-content/src/lib/fields/dot-edit-content-text-area/dot-edit-content-text-area.component.ts`, `core-web/libs/edit-content/src/lib/fields/dot-edit-content-text-area/dot-edit-content-text-area.constants.ts`) [[1]](diffhunk://#diff-a814e235411c98f4b3f57ed83e78c58f49bc2baf95e6dadce63527a7f5122520R1-R18) [[2]](diffhunk://#diff-cfe50a56610dac5036dda6a266fdd9e9942cc455f6fb11d3040f2afee93725b5L1-R49) [[3]](diffhunk://#diff-4260004dce0e7ec19ae73b8c1d189f1ebe5cffe04c7f8775a7db2f39b3126fcdR1-R17).
* Implemented automatic editor mode detection based on content analysis using the `detectEditorType` utility (`core-web/libs/edit-content/src/lib/fields/dot-edit-content-text-area/utils/content-type-detector.util.ts`).

Refactoring of `dot-edit-content-wysiwyg-field` component:

* Replaced the custom language variable selector with the `dot-language-variable-selector` component for consistency and maintainability (`core-web/libs/edit-content/src/lib/fields/dot-edit-content-wysiwyg-field/dot-edit-content-wysiwyg-field.component.html`, `core-web/libs/edit-content/src/lib/fields/dot-edit-content-wysiwyg-field/dot-edit-content-wysiwyg-field.component.scss`) [[1]](diffhunk://#diff-75d0ebdac6be6cdfbdd0be81b738793c3071d7a0badc02ad60f8fa62cc5d563dR1-R2) [[2]](diffhunk://#diff-2b50090eee172a922f7acd68dcd40eb94251976bd1d4c7c3ec98ccd7e705140aR1-R8).
* Removed the unused `ngx-monaco-editor` component from the WYSIWYG field template (`core-web/libs/edit-content/src/lib/fields/dot-edit-content-wysiwyg-field/components/dot-wysiwyg-monaco/dot-wysiwyg-monaco.component.html`).

Supporting style updates:

* Updated styles to ensure proper layout and alignment of the new editor controls and components (`core-web/libs/dotcms-scss/angular/dotcms-theme/components/form/common.scss`, `core-web/libs/edit-content/src/lib/components/dot-edit-content-sidebar/dot-edit-content-sidebar.component.scss`) [[1]](diffhunk://#diff-7d506dfae172a6ef5f0470dee14a3178e9df9fad23e162c8be6605e01a46bc51R70) [[2]](diffhunk://#diff-40fc1f237ad32dfdf6745feec4f4a28133bc91108cc33a3689679984c991b81fR21-R22) [[3]](diffhunk://#diff-40fc1f237ad32dfdf6745feec4f4a28133bc91108cc33a3689679984c991b81fR45).

Refactoring of test cases:

* Simplified the test cases for the WYSIWYG field component by removing unnecessary mocks and dependencies (`core-web/libs/edit-content/src/lib/fields/dot-edit-content-wysiwyg-field/dot-edit-content-wysiwyg-field.component.spec.ts`) [[1]](diffhunk://#diff-b2d7e080973018db72ba464358ae5bc33312adf4b1c174df7e45f1aa76e7d364L2-R2) [[2]](diffhunk://#diff-b2d7e080973018db72ba464358ae5bc33312adf4b1c174df7e45f1aa76e7d364L20-R14) [[3]](diffhunk://#diff-b2d7e080973018db72ba464358ae5bc33312adf4b1c174df7e45f1aa76e7d364L46-L83) [[4]](diffhunk://#diff-b2d7e080973018db72ba464358ae5bc33312adf4b1c174df7e45f1aa76e7d364L95) [[5]](diffhunk://#diff-b2d7e080973018db72ba464358ae5bc33312adf4b1c174df7e45f1aa76e7d364L125) [[6]](diffhunk://#diff-b2d7e080973018db72ba464358ae5bc33312adf4b1c174df7e45f1aa76e7d364L146-L149).

### Checklist
- [x] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  ** updated screenshot **

This PR fixes: #31244

oidacra committed:
feat(edit-content): enhance text area component with Monaco editor support

This commit introduces significant updates to the  component, enabling dual editing modes: plain text and code editing via the Monaco editor. Key changes include:

- Refactored the HTML structure to accommodate a dropdown for selecting the editor type and a language variable selector.
- Implemented a new  for the Monaco editor integration, including custom styling and configuration options.
- Added utility functions for detecting the appropriate editor based on content type.
- Updated SCSS styles for improved layout and consistency across the text area component.

These enhancements aim to provide a more versatile editing experience within the DotCMS environment, allowing users to seamlessly switch between text and code editing modes.

oidacra committed:
style(autocomplete): enhance dropdown and input styles for better consistency

This commit updates the SCSS styles for the autocomplete and dropdown components across various files. Key changes include:

- Set height and minimum height for the autocomplete dropdown and button icons to ensure consistent sizing.
- Updated the language variable selector styles to improve layout and alignment.
- Refactored class names in the HTML structure for better clarity and maintainability.

These enhancements aim to provide a more cohesive user experience within the DotCMS environment.

oidacra committed:
feat(dot-highlight): add highlight pipe and integrate with language variable selector

This commit introduces a new  that highlights search terms within text. Key changes include:

- Implementation of the  to wrap matching text with a highlight span.
- Integration of the highlight pipe into the  component, enhancing the display of variable keys.
- Updates to the component's HTML and SCSS for improved layout and styling consistency.

These enhancements aim to improve user experience by providing visual feedback on search terms within the language variable selector.

oidacra committed:
feat(edit-content): enhance text area and WYSIWYG components with language variable selector and Monaco editor support

This commit introduces several enhancements to the text area and WYSIWYG components, focusing on improved editor functionality and user experience. Key changes include:

- Integration of a language variable selector in both the text area and WYSIWYG components, allowing users to easily select and insert language variables.
- Implementation of a dropdown for selecting between plain text and Monaco editor modes, enhancing the editing experience.
- Updates to component tests to ensure proper functionality of the new features, including editor mode switching and language variable selection.

These improvements aim to provide a more versatile and user-friendly editing environment within the DotCMS platform.

oidacra committed:
fix(edit-content) fix format issues

hmoreras commented on core-web/libs/ui/src/lib/components/dot-language-variable-selector/dot-language-variable-selector.component.scss:
```diff
@@ -0,0 +1,90 @@
+                }
+
+                .dot-language-variable-selector__item-key {
+                    font-size: 12px;
```
rem here ?

oidacra replied:
fixed

nicobytes commented on core-web/libs/ui/src/lib/components/dot-language-variable-selector/dot-language-variable-selector.component.ts:
```diff
@@ -0,0 +1,215 @@
+    /**
+     * Signal to get the autocomplete component.
+     */
+    $autoComplete = viewChild<AutoComplete>(AutoComplete);
```
I think this `viewChild` should be marked as required

nicobytes commented on core-web/libs/ui/src/lib/pipes/dot-highlight/dot-highlight.pipe.ts:
```diff
@@ -0,0 +1,61 @@
+    name: 'dotHighlight',
+    standalone: true
+})
+export class DotHighlightPipe implements PipeTransform {
```
Amazing pipe!

oidacra commented on core-web/libs/edit-content/src/lib/fields/dot-edit-content-text-area/dot-edit-content-text-area.component.html:
```diff
@@ -1,9 +1,31 @@
+                [id]="field.variable"
+                [formControlName]="field.variable"
+                [attr.data-testId]="field.variable"
+                [style]="{
```
added

oidacra commented on core-web/libs/edit-content/src/lib/fields/dot-edit-content-text-area/dot-edit-content-text-area.component.ts:
```diff
@@ -19,6 +55,138 @@ import { DotCMSContentTypeField } from '@dotcms/dotcms-models';
+    /**
+     * Input field DotCMSContentTypeField
+     */
+    $field = input<DotCMSContentTypeField>({} as DotCMSContentTypeField, { alias: 'field' });
```
done.

oidacra committed:
feat(edit-content): enhance text area component with null input handling and add content type detector tests

This commit introduces improvements to the DotEditContentTextAreaComponent by allowing the input field to accept null values, enhancing its robustness. Additionally, a new test suite for the content type detection utility has been added, ensuring accurate identification of editor types based on content syntax. These changes aim to improve the overall functionality and reliability of the editing components within the DotCMS platform.

oidacra committed:
feat(edit-content) fix cr comments
